File: Python Problems/132.PalindromPartition.cpp
class Solution {
public:
    int dp[2001][2001];
    bool isPal(string &s,int i,int j){
        if(i>=j)return true;
        while(i<j){
            if(s[i++]!=s[j--])return false;
        }
        return true;
    }
    int solve(string &s,int i,int j){
        if(i>=j || isPal(s,i,j))return 0;
        if(dp[i][j]!=-1)return dp[i][j];
        int ans=INT_MAX;
        for(int k=i;k<=j-1;k++){
            if(isPal(s,i,k)){
                ans=min(ans,1+solve(s,k+1,j));
            }
        }
        return dp[i][j]=ans;
    }
    int minCut(string s) {
        int n=s.size();
        memset(dp,-1,sizeof(dp));
        return solve(s,0,n-1);
    }
};

File: Python Problems/132.PalindromPartition.py
from sys import maxsize
class Solution:
    def __init__(self) -> None:
        self.it=0
        self.arr=[[]]
    def minCut(self, s: str) -> int:
        
        self.solve(s,0,len(s)-1)
        print( self.arr)
    def solve(self,s,i,j):
        if i>j :return 
        if self.isPalindrome(s,i,j): 
            self.arr[self.it].append(s[i:j+1]) 
            return
        k=i
        while k<=j-1:
            self.solve(s,i,k)
            self.solve(s,k+1,j)
            k=k+1
        self.arr.append([])
        self.it=self.it+1
        return 
    def isPalindrome(self,s,i,j):
        if i>=j :return True
        while i<j: 
            if s[i]!=s[j] : 
                return False
            i=i+1
            j=j-1
        return True

ans= Solution()
ans.minCut('abccb')



